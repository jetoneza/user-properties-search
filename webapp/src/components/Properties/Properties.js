import React from 'react';
import PropTypes from 'prop-types';
import { Grid, Card, Icon, Segment, Loader, Image } from 'semantic-ui-react';
import { imageUrlFaker } from '../../utils/mocks';

function Properties({ list, loading }) {
  if (loading) {
    return (
      <Loader
        inline="centered"
        active
        size="large">
        Loading
      </Loader>
    );
  }

  if (!list || list.length === 0) {
    return <Segment>No results found.</Segment>;
  }

  return (
    <Grid centered stackable columns={3}>
      { list.map((property) => {
        return (
          <Grid.Column key={property.id}>
            <Card color="green">
              <Image src={ imageUrlFaker('house', property.id) } />
              <Card.Content>
                <Card.Header>{ property.street }</Card.Header>
                <Card.Meta>
                  <span>{ property.city }, { property.state }</span>
                </Card.Meta>
                <Card.Description>${ property.rent }</Card.Description>
              </Card.Content>
              { property.user && (
                <Card.Content extra>
                  <a>
                    <Icon name='user' />
                    { property.user.firstName } { property.user.lastName }
                  </a>
                </Card.Content>
              ) }
            </Card>
          </Grid.Column>
        );
      }) }
    </Grid>
  );
}

Properties.propTypes = {
  list: PropTypes.arrayOf(
    PropTypes.shape({
      street: PropTypes.string.isRequired,
      city: PropTypes.string.isRequired,
      state: PropTypes.string.isRequired,
      rent: PropTypes.number.isRequired,
      user: PropTypes.shape({
        firstName: PropTypes.string.isRequired,
        lastName: PropTypes.string.isRequired,
      }),
    })
  ).isRequired,
}

export default Properties;
