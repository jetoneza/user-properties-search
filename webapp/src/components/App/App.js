import React from 'react';
import PropTypes from 'prop-types';
import { graphql } from 'react-apollo';
import gql from 'graphql-tag';
import { Segment, Container } from 'semantic-ui-react';

import SearchInput from '../SearchInput';
import Properties from '../Properties';

export class App extends React.Component {
  static propTypes = {
    loading: PropTypes.bool.isRequired,
    properties: PropTypes.array.isRequired,
    search: PropTypes.func.isRequired,
  }

  state = {
    query: '',
    page: 1,
  }

  handleSearchSubmit = (query) => {
    this.setState({ query });

    this.props.search({ query, page: this.state.page });
  }

  render () {
    const { properties, loading } = this.props;

    return (
      <Segment vertical style={{ padding: '8em 0' }}>
        <Container>
          <h1>Properties Search</h1>
          <SearchInput onSubmit={this.handleSearchSubmit}/>
          <Properties properties={properties} loading={loading}/>
        </Container>
      </Segment>
    );
  }
}

const query = gql`
  query($query: String, $page: Int) {
    search(query: $query, page: $page) {
      id
      street
      city
      state
      zip
      rent
      user {
        id
        firstName
        lastName
      }
    }
  }
`;

export default graphql(query, {
  options: (props) => ({
    variables: {
      query: '',
      page: 1,
    }
  }),
  props: ({ data }) => ({
    loading: data.loading,
    properties: data.search || [],
    search: data.refetch,
  }),
})(App);
