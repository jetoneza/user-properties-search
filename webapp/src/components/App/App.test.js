import React from 'react';
import renderer from 'react-test-renderer';

import { shallow } from 'enzyme';

import { App } from './App';
import Properties from '../Properties';

describe('<App /> test', () => {
  const props = {
    loading: false,
    properties: {
      results: [],
    },
    search: jest.fn(),
  }

  test('Should render properties list in app', () => {
    const wrapper = shallow(<App {...props}/>);

    expect(wrapper.find(Properties).length).toEqual(1);
  });

  test('Should render component', () => {
    const component = <App {...props}/>
    shallow(component);

    const renderedComponent = renderer.create(component);
    const tree = renderedComponent.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
