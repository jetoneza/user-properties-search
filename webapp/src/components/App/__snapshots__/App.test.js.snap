// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> test Should render component 1`] = `
<div
  className="ui vertical segment"
  style={
    Object {
      "padding": "8em 0",
    }
  }
>
  <div
    className="ui container"
  >
    <h1>
      Properties Search
    </h1>
    <div
      className="ui centered grid"
    >
      <div
        className="sixteen wide column"
      >
        <form
          action={undefined}
          className="ui form"
          onSubmit={[Function]}
        >
          <div
            className="ui action input"
          >
            <input
              disabled={undefined}
              onChange={[Function]}
              placeholder="Search"
              tabIndex={undefined}
              type="text"
              value=""
            />
            <button
              className="ui icon button"
              disabled={undefined}
              onClick={[Function]}
              role="button"
              tabIndex={undefined}
            >
              <i
                aria-hidden="true"
                className="search icon"
              />
            </button>
          </div>
        </form>
      </div>
    </div>
    <div
      className="ui segment"
    >
      No results found.
    </div>
  </div>
</div>
`;
